nupic/torch/functions/k_winners.py

```diff
@@ -21,229 +21,187 @@
-          gradients for the backward pass.
-
-        :param x:
-          Current activity of each unit.
-
-        :param duty_cycles:
-          The averaged duty cycle of each unit.
-
-        :param k:
-          The activity of the top k units across the channels will be allowed to
-          remain, the rest are set to zero.
-
-        :param boost_strength:
-          A boost strength of 0.0 has no effect on x.
-
-        :return:
-             A tensor representing the activity of x after k-winner take all.
-        """
-        boosted = boost_activations(x, duty_cycles, boost_strength)
-
-        # Select top K channels from the boosted values
-        topk, indices = boosted.topk(k=k, dim=1)
-        res = torch.zeros_like(x)
-        res.scatter_(1, indices, x.gather(1, indices))
-
-        ctx.save_for_backward(indices)
-        return res
-
-    @staticmethod
-    def backward(ctx, grad_output):
-        """
-        In the backward pass, we set the gradient to 1 for the winning units,
-        and 0 for the others.
-        """
-        indices, = ctx.saved_tensors
-        grad_x = torch.zeros_like(grad_output, requires_grad=False)
-        grad_x.scatter_(1, indices, grad_output.gather(1, indices))
-        return grad_x, None, None, None, None
+    If local=True, k-winners are chosen independently for each location. For
+    Conv2d inputs (batch, channel, H, W), the top k channels are selected
+    locally for each of the H X W locations. If there is a tie for the kth
+    highest boosted value, there will be more than k winners.
+
+    The boost strength is used to compute a boost factor for each unit
+    represented in x. These factors are used to increase the impact of each unit
+    to improve their chances of being chosen. This encourages participation of
+    more columns in the learning process. See :meth:`kwinners` for more details.
+
+    :param x:
+      Current activity of each unit.
+
+    :param duty_cycles:
+      The averaged duty cycle of each unit.
+
+    :param k:
+      The activity of the top k units across the channels will be allowed to
+      remain, the rest are set to zero.
+
+    :param boost_strength:
+      A boost strength of 0.0 has no effect on x.
+
+    :param local:
+      Whether or not to choose the k-winners locally (across the channels at
+      each location) or globally (across the whole input and across all
+      channels).
+
+    :param break_ties:
+      Whether to use a strict k-winners. Using break_ties=False is faster but
+      may occasionally result in more than k active units.
+
+    :param relu:
+      Whether to simulate the effect of applying ReLU before KWinners.
+
+    :param inplace:
+      Whether to modify x in place
+
+    :return:
+         A tensor representing the activity of x after k-winner take all.
+    """
+    if k == 0:
+        return torch.zeros_like(x)
+
+    boosted = boost_activations(x, duty_cycles, boost_strength)
+
+    if break_ties:
```

Comment:
All these options are determined once when the network is specified, but they are checked for each forward pass. I wonder how much these conditionals slow things down on a SIMD GPU? For example, what if we had a version of this that assumed a specific set of options and had no if statements? Do you have any thoughts on whether that would have any appreciable effect on timing?

Reply:
Because PyTorch is asynchronous, I think it's fine. Yes, the Python thread will take slightly longer than needed, but the important thing is that it keeps the GPU purring. I think removing the overhead of these checks would have zero improvement (not just negligible; truly zero) because at this point, the Python code is way ahead of the GPU.

This is one of PyTorch's key innovations: the fact that we can run models in a slow language like Python and pay almost zero cost for it. (by using an asynchronous approach where the CPU is always multiple steps ahead of the GPU)

Reply:
Interesting, thanks. Did not know that.